src/core/net/nd_agent.cpp

```diff
@@ -0,0 +1,133 @@
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements IPv6 Neighbor Discovery Agent.
+ */
+
+#include "nd_agent.hpp"
+
+#if OPENTHREAD_CONFIG_NEIGHBOR_DISCOVERY_AGENT_ENABLE
+
+#include "common/locator-getters.hpp"
+#include "thread/lowpan.hpp"
+#include "thread/mle_router.hpp"
+#include "thread/network_data_leader.hpp"
+#include "thread/thread_netif.hpp"
+
+namespace ot {
+namespace NeighborDiscovery {
+
+void Agent::UpdateService(void)
+{
+    Error                           error;
+    uint16_t                        rloc16 = Get<Mle::MleRouter>().GetRloc16();
+    NetworkData::Iterator           iterator;
+    NetworkData::OnMeshPrefixConfig config;
+    Lowpan::Context                 lowpanContext;
+
+    if (mAloc.mNext != &mAloc)
+    {
+        uint8_t contextId = static_cast<uint8_t>(mAloc.GetAddress().GetIid().GetLocator() -
+                                                 Mle::kAloc16NeighborDiscoveryAgentStart + 1);
+        bool    found     = false;
+
+        iterator = NetworkData::kIteratorInit;
+
+        while (Get<NetworkData::Leader>().GetNextOnMeshPrefix(iterator, rloc16, config) == kErrorNone)
+        {
+            if (!config.mNdDns)
+            {
+                continue;
+            }
+
+            error = Get<NetworkData::Leader>().GetContext(static_cast<const Ip6::Address &>(config.mPrefix.mPrefix),
+                                                          lowpanContext);
+
+            if ((error != kErrorNone) || (lowpanContext.mContextId != contextId))
+            {
+                continue;
+            }
+
+            found = true;
+            break;
+        }
+
+        if (!found)
+        {
+            Get<ThreadNetif>().RemoveUnicastAddress(mAloc);
+            mAloc.mNext = &mAloc;
+        }
+    }
+
+    VerifyOrExit(mAloc.mNext == &mAloc);
+
+    iterator = NetworkData::kIteratorInit;
+
+    while (Get<NetworkData::Leader>().GetNextOnMeshPrefix(iterator, rloc16, config) == kErrorNone)
+    {
+        if (!config.mNdDns)
+        {
+            continue;
+        }
+
+        error = Get<NetworkData::Leader>().GetContext(static_cast<const Ip6::Address &>(config.mPrefix.mPrefix),
+                                                      lowpanContext);
+
+        if (error == kErrorNone)
+        {
+            uint16_t rloc = Mle::kAloc16NeighborDiscoveryAgentStart + lowpanContext.mContextId - 1;
+
+            mAloc.InitAsThreadOrigin();
+            mAloc.GetAddress().SetToAnycastLocator(Get<Mle::MleRouter>().GetMeshLocalPrefix(), rloc);
+            Get<ThreadNetif>().AddUnicastAddress(mAloc);
+            ExitNow();
+        }
+    }
+
+exit:
+    return;
+}
+
+void Agent::ApplyMeshLocalPrefix(void)
+{
+    VerifyOrExit(mAloc.mNext != &mAloc);
```

Comment:
```suggestion
    VerifyOrExit(Get<ThreadNetif>().HasUnicastAddress(mAloc));
```

Implement `HasUnicastAddress` in ThreadNetif by searching `mUnicastAddresses` list?

The current code depends on the how `LinkedList`  manages the `mNext` pointer.

Reply:
I think this model may be fine here since `Agent` itself is owner of the `mAloc` variable.

We know that `LinkedList` class chains the entries together and always end with `nullptr` (since that's how we iterate through the list), so having an entry point back to itself is a good way to know that the entry is not in a linked-list.

Reply:
> We know that `LinkedList` class chains the entries together and always end with `nullptr` (since that's how we iterate through the list), so having an entry point back to itself is a good way to know that the entry is not in a linked-list.

An alternative implementation of `LinkedList` is let the laster element's `mNext` pointer points to the first element. In such case, every element of a linked list must have `mNext != nullptr`, and we should use `mAloc.mNext != nullptr` to check if `mAloc` is in use. 
That's why I think the current code depends on how `LinkedList` is implemented.
But this seems to be a minor issue. I will resolve.